File: LeetCode/remote/permutations.cs
// https://leetcode.com/problems/permutations/
//
//  Given a collection of numbers, return all possible permutations.
//
//  For example,
//  [1,2,3] have the following permutations:
//  [1,2,3], [1,3,2], [2,1,3], [2,3,1], [3,1,2], and [3,2,1]. 
//
// 
// Submission Details
// 25 / 25 test cases passed.
//  Status: Accepted
//  Runtime: 556 ms
//      
//      Submitted: 10 hours, 52 minutes ago
//

using System;
using System.Collections.Generic;

public class Solution {
    public IList<IList<int>> Permute(int[] nums) {
        var result = new List<IList<int>>();
        if (nums.Length < 2)
        {
            result.Add(nums.ToList());
            return result;
        }

        var number = nums.First();
        foreach (var prev in Permute(nums.Skip(1).ToArray()).ToList())
        {
            result.AddRange(Enumerable
                             .Range(0, nums.Length)
                             .Select(x => {
                                var newList = new List<int>(prev);
                                newList.Insert(x, number);
                                return newList;                                
                             }));
        }

        return result;
    }

    static void Main()
    {
        Console.WriteLine(String.Join(Environment.NewLine), Permute(new [] { 1, 2, 3 }).Select(x => String.Join(", ", x)));
    }
}

//  https://leetcode.com/submissions/detail/58538909/
//
//  Submission Details
//  25 / 25 test cases passed.
//      Status: Accepted
//      Runtime: 564 ms
//          
//          Submitted: 0 minutes ago
//
public class Solution {
    public IList<IList<int>> Permute(int[] nums) {
        IList<IList<int>> result = new List<IList<int>>();
        if (nums.Length == 1)
        {
            result.Add(new List<int> { nums[0]});
            return result;
        }
        
        foreach (var next in nums
                             .SelectMany(
                                 x => Permute(nums.Except(new List<int> { x }).ToArray()),
                                 (x, y) => {
                                     return new List<int> { x }.Concat(y).ToList();
                                 }).ToList())
        {
            result.Add(next);                             
        }
        
        return result;
    }
}

//  Do the evolution
//  https://leetcode.com/submissions/detail/66209409/
//
//  Submission Details
//  25 / 25 test cases passed.
//      Status: Accepted
//      Runtime: 472 ms
//          
//          Submitted: 0 minutes ago
public class Solution {
    public IList<IList<int>> Permute(int[] nums) {
        if (nums.Length < 2)
        {
            return new List<IList<int>> { nums.ToList() };
        }
        
        return nums.SelectMany(
            x => Permute(nums.Except(new int[] { x }).ToArray()),
            (x, y) => new List<int> { x }.Concat(y).ToList())
            .ToList<IList<int>>();
    }
}

File: LeetCode/remote/permutations.rb
#   https://leetcode.com/problems/permutations/
#
#    Given a collection of distinct numbers, return all possible permutations. 
#
#    Oh how I fucking hate Ruby flattening \ array ambiguities
#
#    https://leetcode.com/submissions/detail/53446432/
#
#
#    Submission Details
#    25 / 25 test cases passed.
#       Status: Accepted
#       Runtime: 96 ms
#           
#           Submitted: 0 minutes ago

# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums)
    return [[]] if !nums.any?
    return [nums] if nums.size == 1
    result = []
    nums.each { |x|
        permute(nums - [x])
        .map { |y|
            y.unshift(x)
        }
        .each { |y|
            result << y
        }
    }
    
    return result
end

#   Slightly better solution
#   https://leetcode.com/submissions/detail/55270226/
#
#   Submission Details
#   25 / 25 test cases passed.
#       Status: Accepted
#       Runtime: 116 ms
#           
#           Submitted: 0 minutes ago
def permute(nums)
    return [[nums.first]] if nums.size == 1
    nums.inject([]) { |acc, x|
        acc << permute(nums - [x]).map { |y|
            [x] + y
        }
    }.flatten(1)
end

#   https://leetcode.com/submissions/detail/55270600/
#   
#   Submission Details
#   25 / 25 test cases passed.
#       Status: Accepted
#       Runtime: 104 ms
#           
#           Submitted: 0 minutes ago
def permute(nums)
    result = []
    permutations(nums, 0, result)
    result
end

def permutations(nums, position, result)
    result << nums.clone if position == nums.size
    (position...nums.size).each { |i|
        nums[position], nums[i] = nums[i], nums[position]
        permutations(nums, position + 1, result)
        nums[position], nums[i] = nums[i], nums[position]
    }
end

